keps/sig-api-machinery/1965-kube-apiserver-identity/README.md

```diff
@@ -65,14 +80,24 @@ load balancer for the cluster, where the advertise IP address is set to the IP
 address of the load balancer, all three kube-apiservers will have the same
 advertise IP address.
 
+### Goals
+
+* Provide a mechanism in which controllers can uniquely identify kube-apiserver's in a cluster.
+
+### Non-Goals
+
+* improving the availability of kube-apiserver
+
 ## Proposal
 
 We will use “hostname+PID+random suffix (e.g. 6 base58 digits)” as the ID.
 
-Similar to the [node heartbeat](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/0009-node-heartbeat.md),
+Similar to the [node heartbeats](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/589-efficient-node-heartbeats),
 a kube-apiserver will store its ID in a Lease object. All kube-apiserver Leases
-will be stored in a special namespace “kube-apiserver-lease”. A controller will
-garbage collect expired Leases.
+will be stored in a special namespace `kube-apiserver-lease`. The Lease creation
+and heart beat will be managed by a controller that is started in kube-apiserver's
+post startup hook. A separate controller in kube-controller-manager will be responsible
```

Comment:
I'm not sure I agree with putting that in controller manager, apiserver gets updated first and the new version of apiserver needs to get to healthy without assistance from controller-manager.

Reply:
The controller in kube-controller-manager is only for garbage collecting stale leases (https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/storageversiongc/gc_controller.go#L17), so I don't _think_ it should impact health of newly started apiservers, but I have not fully tested this myself.

Reply:
If it's not running at all, leases will build up over time and be confusing at best and affect the performance of consumers at worst, right? It feels like bad karma but I guess I can't think of anything disastrous...